### PR TITLE
subsys: modem: optimize memory footprint

### PR DESCRIPTION
**Pull request by ndrs-pst**
Series of commits to optimize memory footprint in `modem_cmux.c`
- modem: pipe: make `modem_pipe_api` a pointer to constant
- modem: cmux: apply `static const` to reduce RAM usage
- modem: cmux: optimize `modem_cmux_frame` structure padding

### CODE REVIEW

**Comment by ndrs-pst** [comment]
Revise the 3rd commit: Instead of reordering the `type` field, downsizing `dlci_address` field from `uint16_t` to `uint8_t` makes more sense.

**Review by bjarki-andreasen — DISMISSED** [review]
Love it :)

File: include/zephyr/modem/pipe.h [hunk 1]
```diff
@@ -77,7 +77,7 @@ struct modem_pipe {
  * @param data Pipe data to bind to pipe instance
  * @param api Pipe API implementation to bind to pipe instance
  */
-void modem_pipe_init(struct modem_pipe *pipe, void *data, struct modem_pipe_api *api);
+void modem_pipe_init(struct modem_pipe *pipe, void *data, const struct modem_pipe_api *api);
```
Comment (tomi-font) [inline include/zephyr/modem/pipe.h:80]: If you want to go full circle with this, you could also add `static const` to instantiations of `struct modem_pipe_api` (in the CMUX module and the various backends).

**Review by tomi-font — DISMISSED** [review]
Nice!

File: include/zephyr/modem/pipe.h (hunk 1 above)
Comment (ndrs-pst) [reply include/zephyr/modem/pipe.h:80]: OK, got it.

**Review by tomi-font — APPROVED** [review]

**Review by bjarki-andreasen — APPROVED** [review]

**Review by rerickson1 — APPROVED** [review]

**Pull request merged**